Cleanup Buffer tests

hughbe opened the pull request:
- Take advantage of xunit
- Add some missing tests

Contributes to #6286
Cherry picked from #5490

hughbe commented:
Test Innerloop CentOS7.1 Release Build and Test

stephentoub commented on src/System.Runtime/tests/System/Buffer.cs:
```diff
@@ -4,111 +4,211 @@
+        {
+            Assert.Equal(sourceArray[i], destinationArray[i + destinationIndexOffset - sourceIndexOffset]);
+        }
+    }
```
Could you add loops verifying that before and after the range in the destinationArray remain 0?

stephentoub commented on src/System.Runtime/tests/System/Buffer.cs:
```diff
@@ -4,111 +4,211 @@
+            sourceArray[i] = i;
+        }
+
+        var destinationArray = new int[30000];
```
Could you put the 25000 and the 30000 into named consts, that are then used here and below?

stephentoub replied:
Then once everything is in the InlineData, there are some other inputs that would be good.  MemoryCopy has special cases for each length up to 16, for example.  It has special cases for when there are various remaining counts after it's copied the bulk of the data.  Etc.  It'd be good to look at the code coverage numbers for Memmove (used by MemoryCopy) and ensure each of those special cases are being covered... I expect that can be done just by adding additional InlineDatas here.

stephentoub commented on src/System.Runtime/tests/System/Buffer.cs:
```diff
@@ -4,111 +4,211 @@
 
     [Fact]
-    public static void TestGetByteAndSetByte()
+    public static unsafe void TestMemoryCopy_OverlappingBuffersSmallCopy()
```
Isn't this test identical to the previous one except for a few different numbers? Rather than duplicating it, could it be a theory with different inputs?

stephentoub commented on src/System.Runtime/tests/System/Buffer.cs:
```diff
@@ -4,111 +4,211 @@
+        yield return new object[] { new byte[] { 0x1a, 0x2b, 0x3c, 0x4d }, 0, new byte[] { 0x6f, 0x6f, 0x6f, 0x6f, 0x6f, 0x6f }, 1, 4, new byte[] { 0x6f, 0x1a, 0x2b, 0x3c, 0x4d, 0x6f } };
+
+        var dst = new byte[] { 0x1a, 0x2b, 0x3c, 0x4d, 0x5e };
+        yield return new object[] { dst, 1, dst, 2, 2, new byte[] { 0x1a, 0x2b, 0x2b, 0x3c, 0x5e } };
```
These are long lines with multiple nested arrays.  At the very least, I'd suggest splitting across multiple lines with comments to indicate what each piece is, e.g. instead of:
```
yield return new object[] { new byte[] { 0x1a, 0x2b, 0x3c, 0x4d }, 0, new byte[] { 0x6f, 0x6f, 0x6f, 0x6f, 0x6f, 0x6f }, 1, 4, new byte[] { 0x6f, 0x1a, 0x2b, 0x3c, 0x4d, 0x6f } };
```
do:
```
yield return new object[] { 
    new byte[] { 0x1a, 0x2b, 0x3c, 0x4d }, // src
    0, // srcOffset
    new byte[] { 0x6f, 0x6f, 0x6f, 0x6f, 0x6f, 0x6f }, // dst
    1, // dstOffset
    4, // count
    new byte[] { 0x6f, 0x1a, 0x2b, 0x3c, 0x4d, 0x6f } // expected
};
```
But, given that there are only two inputs here, it might be better to just have a single [Fact] that contains these... there's more ceremony around the enumerable, yield return, etc. than there is in just duplicating the Buffer.BlockCopy and Assert.Equal calls.

stephentoub commented:
Some comments, otherwise LGTM.

hughbe commented:
Thanks, I addressed your comments - unfortunately I couldn't add all the extra tests you asked for, as `sizeof(int) == 4` so I got strange results if `sourceBytesToCopy` was not a value of 4.

If we want to test all inputs up to 16, I think we need to use a different non-integer primitive type

stephentoub commented:
> If we want to test all inputs up to 16, I think we need to use a different non-integer primitive type

Yes, ```byte``` will be the easiest way.

hughbe commented:
> Yes, byte will be the easiest way.

Done

stephentoub commented:
Thanks. LGTM.
